apollo-router/src/plugins/authentication/jwks.rs

```diff
@@ -256,3 +273,419 @@ impl Iterator for Iter<'_> {
+    keys: Vec<(Option<String>, Jwk)>,
+    criteria: JWTCriteria,
+) -> Result<(Option<String>, TokenData<serde_json::Value>), (AuthenticationError, StatusCode)> {
+    let mut error = None;
+    for (issuer, jwk) in keys.into_iter() {
+        let decoding_key = match DecodingKey::from_jwk(&jwk) {
+            Ok(k) => k,
+            Err(e) => {
+                error = Some((
+                    AuthenticationError::CannotCreateDecodingKey(e),
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                ));
+                continue;
+            }
+        };
+
+        let key_algorithm = match jwk.common.key_algorithm {
+            Some(a) => a,
+            None => {
+                error = Some((
+                    AuthenticationError::JWKHasNoAlgorithm,
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                ));
+                continue;
+            }
+        };
+
+        let algorithm = match convert_key_algorithm(key_algorithm) {
+            Some(a) => a,
+            None => {
+                error = Some((
+                    AuthenticationError::UnsupportedKeyAlgorithm(key_algorithm),
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                ));
+                continue;
+            }
+        };
+
+        let mut validation = Validation::new(algorithm);
+        validation.validate_nbf = true;
+        // if set to true, it will reject tokens containing an `aud` claim if the validation does not specify an audience
+        // we don't validate audience yet, so this is deactivated
+        validation.validate_aud = false;
+
+        match decode::<serde_json::Value>(jwt, &decoding_key, &validation) {
+            Ok(v) => return Ok((issuer, v)),
+            Err(e) => {
+                tracing::trace!("JWT decoding failed with error `{e}`");
+                error = Some((
+                    AuthenticationError::CannotDecodeJWT(e),
+                    StatusCode::UNAUTHORIZED,
+                ));
+            }
+        };
+    }
+
+    match error {
+        Some(e) => Err(e),
+        None => {
+            // We can't find a key to process this JWT.
+            Err((
+                criteria.kid.map_or_else(
+                    || AuthenticationError::CannotFindSuitableKey(criteria.alg, None),
+                    AuthenticationError::CannotFindKID,
+                ),
+                StatusCode::UNAUTHORIZED,
+            ))
+        }
+    }
+}
+
+pub(crate) fn jwt_expires_in(context: &Context) -> Duration {
+    let claims = context
+        .get(APOLLO_AUTHENTICATION_JWT_CLAIMS)
+        .map_err(|err| tracing::error!("could not read JWT claims: {err}"))
+        .ok()
+        .flatten();
+    let ts_opt = claims.as_ref().and_then(|x: &serde_json::Value| {
+        if !x.is_object() {
+            tracing::error!("JWT claims should be an object");
+            return None;
+        }
+        let claims = x.as_object().expect("claims should be an object");
```

Comment:
As we try as much as possible to get rid of `expect` and `unwrap` in our code, maybe that could be worth it to use pattern matching here instead of an expect ?

Reply:
I'll rewrite this as a `match`